File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return tra(Long.MIN_VALUE,Long.MAX_VALUE,root);
    }
    public boolean tra(long l,long u,TreeNode r){
        if(r==null){
            return true;
        }
        else if(r.val<=l || r.val>=u){
            return false;
        }
        else{
            return tra(l,r.val,r.left) && tra(r.val,u,r.right);
        }
    }
}

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        INF = sys.maxsize
        
        def tra(l,u,r):
            if not r:
                return True
            elif r.val>=u or r.val<=l:
                return False 
            else:
                return tra(l,r.val,r.left) and tra(r.val,u,r.right)
            
        return tra(-INF,INF,root)
        